iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/dataregion/historical/PipeHistoricalDataRegionTsFileAndDeletionExtractor.java

```diff
@@ -382,113 +507,33 @@ public synchronized void start() {
-                        // PIPE_MIN_FLUSH_INTERVAL_IN_MS. We simply ignore them.
-                        !resource.isClosed()
-                            || mayTsFileContainUnprocessedData(resource)
-                                && isTsFileResourceOverlappedWithTimeRange(resource)
-                                && isTsFileGeneratedAfterExtractionTimeLowerBound(resource)
-                                && mayTsFileResourceOverlappedWithPattern(resource))
-                .collect(Collectors.toList());
-        resourceList.addAll(unsequenceTsFileResources);
-
-        resourceList.forEach(
-            resource -> {
-              // Pin the resource, in case the file is removed by compaction or anything.
-              // Will unpin it after the PipeTsFileInsertionEvent is created and pinned.
-              try {
-                PipeDataNodeResourceManager.tsfile()
-                    .pinTsFileResource(resource, shouldTransferModFile);
-              } catch (final IOException e) {
-                LOGGER.warn("Pipe: failed to pin TsFileResource {}", resource.getTsFilePath());
-              }
-            });
-
-        resourceList.sort(
-            (o1, o2) ->
-                startIndex instanceof TimeWindowStateProgressIndex
-                    ? Long.compare(o1.getFileStartTime(), o2.getFileStartTime())
-                    : o1.getMaxProgressIndex().topologicalCompareTo(o2.getMaxProgressIndex()));
-        pendingQueue = new ArrayDeque<>(resourceList);
-
-        LOGGER.info(
-            "Pipe {}@{}: finish to extract historical TsFile, extracted sequence file count {}/{}, "
-                + "extracted unsequence file count {}/{}, extracted file count {}/{}, took {} ms",
-            pipeName,
-            dataRegionId,
-            sequenceTsFileResources.size(),
-            originalSequenceTsFileCount,
-            unsequenceTsFileResources.size(),
-            originalUnsequenceTsFileCount,
-            resourceList.size(),
-            originalSequenceTsFileCount + originalUnsequenceTsFileCount,
-            System.currentTimeMillis() - startHistoricalExtractionTime);
-      } finally {
-        tsFileManager.readUnlock();
-      }
+      List<PersistentResource> resourceList = new ArrayList<>();
+      // Extract deletions
+      Optional.ofNullable(DeletionResourceManager.getInstance(String.valueOf(dataRegionId)))
+          .ifPresent(mgr -> extractDeletions(mgr, resourceList));
+
+      // Flush TsFiles
+      final long startHistoricalExtractionTime = System.currentTimeMillis();
+      flushTsFilesForExtraction(dataRegion, startHistoricalExtractionTime);
+      // Extract TsFiles
+      extractTsFiles(dataRegion, startHistoricalExtractionTime, resourceList);
+
+      // Sort tsFileResource and deletionResource
+      resourceList.sort(
+          (o1, o2) -> o1.getProgressIndex().topologicalCompareTo(o2.getProgressIndex()));
+      pendingQueue.addAll(resourceList);
```

Comment:
Maybe use "new ArrayDeque<>(resourceList);"? The pending queue is null here...

Reply:
fixed